CARDS-500: As a user, I can access statistics on a dashboard

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-508). 

This also includes a new thing onto the QuestionComponentManager that will allow forms to display Reference inputs to other Questions. At the moment this is just used within the Admin Statistics page.

**Test data**: [statisticsTest.csv](https://github.com/data-team-uhn/cards/files/7137731/statisticsTest.csv)
Upload with `curl -u admin:admin "localhost:8080/Forms/" -X POST -F ":data=@statisticsTest.csv" -F":questionnaire=/Questionnaires/Patient information" -F ":subjectType=/SubjectTypes/Patient"`

[statisticsTestTumor.csv](https://github.com/data-team-uhn/cards/files/7137733/statisticsTestTumor.csv)
Upload with `curl -u admin:admin "localhost:8080/Forms/" -X POST -F ":data=@statisticsTestTumor.csv" -F":questionnaire=/Questionnaires/Tumors" -F ":subjectType=/SubjectTypes/Patient" -F ":subjectType=/SubjectTypes/Patient/Tumor"`

Both should be uploaded to test statistics.

anusha-dey committed:
LFS-507, LFS-508

create statistic pages in admin dashboard and sidebar

OrdiNeu committed:
LFS-500: As a user, I can access statistics on a dashboard
Add StatisticQueryServlet.java, which will enable querying the counts of
data for a given lfs:Statistic

anusha-dey committed:
LFS-500: As a user, I can access statistics on a dashboard

sdumitriu committed:
[misc] Update code to work with the new uuid library version

sdumitriu committed:
[misc] Removed unused dependency

OrdiNeu committed:
LFS-500 / LFS-507: New entry in the side menu: Statistics

OrdiNeu committed:
LFS-500 / LFS-507: New entry in the side menu: Statistics
Another emergency bugfix

OrdiNeu committed:
LFS-500 / LFS-507: New entry in the side menu: Statistics

Add support for fields that show up in more than one form

acrowthe committed:
LFS-500: As a user, I can access statistics on a dashboard
Fix issues that arose from rebasing onto latest dev

marta- committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics - fixed admin entry label
LFS-500 / LFS-507: New entry in the side menu: Statistics - fixed sidebar entry label

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics

Update the admin page according to the new specs.
Fix a bug that was preventing ReferenceInputs from displaying their
values

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics
Fix the new statistics submission according to the new
ReferenceInput.jsx

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics

Fix the edit button, remove debugging output

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics

Improve error handling with missing mandatory properties

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics

Allow for numeric-only statistics, and also allow for variables to refer
to multiple questions.

marta- commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,309 @@
+
+  return (
+    <form action='/Statistics' method='POST' onSubmit={saveData}>
+      <Dialog disablePortal open={open} onClose={onClose}>
```
The dialog is too narrow. Use the [ResponsiveDialog](https://github.com/ccmbioinfo/lfs/blob/dev/modules/commons/src/main/frontend/src/components/ResponsiveDialog.jsx) component instead.

marta- commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,309 @@
+    requestData.append('jcr:primaryType', 'lfs:Statistic');
+
+    // Verify that the name, xVar, and yVar variables have been filled out
+    let mandatoryFields = ["name", "xVar", "yVar"];
```
Filelds should be specified as mandatory in the JSON.

OrdiNeu replied:
Implemented in `db66db05` :heavy_check_mark:

sashaandjic commented:
Split var is not shown:
![image](https://user-images.githubusercontent.com/23243244/123206035-f1a9a600-d488-11eb-8d86-45e7294c9508.png)

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics
Use ResponsiveDialog instead of Dialog

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics
Fix an issue that prevented the edit command from saving properly

sashaandjic commented:
Created these 2:
![image](https://user-images.githubusercontent.com/23243244/124874104-4d9a2180-df95-11eb-9f9c-4fa198d50868.png)

When I click on the Statistic, I get this:
![image](https://user-images.githubusercontent.com/23243244/124874045-3c511500-df95-11eb-8c8d-5b8e43c70b78.png)

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics
Move the mandatory field to the JSON instead of being hard-coded

OrdiNeu commented:
@sashaandjic 
> When I click on the Statistic, I get this:
> ![image](https://user-images.githubusercontent.com/23243244/124874045-3c511500-df95-11eb-8c8d-5b8e43c70b78.png)

Should be fixed on. The statistics page doesn't look that great, but it won't crash.

OrdiNeu committed:
LFS-500 / LFS-508: New entry in the admin screen: Statistics
Fix a wrong merge

sashaandjic commented:
I am unable to make a build
![image](https://user-images.githubusercontent.com/23243244/126404464-88b31b99-fb61-4fa8-9195-7ced10c7e7ae.png)

OrdiNeu commented:
@sashaandjic 
> I am unable to make a build (mvn clean install)

Sorry about that, I forgot to force-push my changes.

sashaandjic reviewed (approved):
looking good so far

sashaandjic commented:
Is `Split var` mandatory? If so, prevent saving **Stats** if there is no value selected there.
Whenever I omit `Split var` I get this when I try to load **Stats** from the Dashboard.

![image](https://user-images.githubusercontent.com/23243244/126603764-d5e9dd57-5e46-4030-a3bf-627cef417692.png)
------

Either change "xVar" and "yVar" to "X var" and "Y var" in the error message
![image](https://user-images.githubusercontent.com/23243244/126604552-21c1246a-b9d0-4fec-8a01-6e3ba58ed7c8.png)
------
or use "Please fill out this field." as in the Questionnaire designer:
-----
![image](https://user-images.githubusercontent.com/23243244/126604422-eb6e58db-5920-4cdf-aa99-091680ed4b89.png)

sashaandjic commented:
Please apply the changes made in [Clarify the source of filter variables](https://phenotips.atlassian.net/browse/CARDS-360), so the user will know where the question is coming from.

OrdiNeu committed:
[wip]

OrdiNeu committed:
LFS-508: New entry in the admin screen: Statistics
Switch to using the per-questionnaire question selection in
adminstatistics

OrdiNeu commented:
@sashaandjic 
> Please apply the changes made in [Clarify the source of filter variables](https://phenotips.atlassian.net/browse/CARDS-360) to the `X var` selection, so the user will know where the question is coming from.

Fixed in `748741f` :heavy_check_mark:

OrdiNeu committed:
CARDS-1132: New charting component used by every chart widget
Add the line graph, and break out the statistics into its own component

OrdiNeu committed:
CARDS-1132 / CARDS-509: Sample stats for LFS

OrdiNeu committed:
CARDS-1132: Display the configured statistics on the stats dashboard
Fix a quick bug on the user statistics page

OrdiNeu committed:
CARDS-1261: Administration' s Statistics: The Y and Split var and should be depended on the selection made in X var field

Make the fields context to allow cross-dependencies between fields

OrdiNeu committed:
CARDS-1261: Administration' s Statistics: The Y and Split var and should
be depended on the selection made in X var field

OrdiNeu committed:
CARDS-1261: Administration' s Statistics: The Y and Split var and should
be depended on the selection made in X var field

Prevent users from selecting an invalid option after finding
restrictions

OrdiNeu commented:
Attached is the .csv I was using for testing (though it only does the `Age at 1st malignant diagnosis` field, so a custom statistic needs to be made to track it)
[statisticsTest.csv](https://github.com/data-team-uhn/cards/files/7092015/statisticsTest.csv)

marta- commented on modules/statistics/src/main/resources/SLING-INF/content/Extensions/AdminDashboard/AdminStatistics.json:
```diff
@@ -0,0 +1,9 @@
+    "cards:extensionPointId": "cards/coreUI/admindashboard/entry",
+    "cards:extensionName": "Statistics",
+    "cards:targetURL": "/content.html/admin/Statistics",
+    "cards:icon": "asset:cards-statistics.subjectTypeIcon.js",
```
Wrong icon. Should be the `BarChart`  Material UI icon.

marta- replied:
```suggestion
    "cards:icon": "asset:cards-statistics.statsIcon.js",
```

marta- commented on modules/statistics/src/main/resources/SLING-INF/content/Extensions/Sidebar/UserStatistics.json:
```diff
@@ -0,0 +1,8 @@
+    "cards:extensionPointId": "cards/coreUI/sidebar/entry",
+    "cards:extensionName": "Statistics",
+    "cards:targetURL": "/content.html/Statistics/User",
+    "cards:icon": "asset:cards-statistics.subjectsIcon.js",
```
Wrong icon. Should be the `BarChart`  Material UI icon.

marta- replied:
```suggestion
    "cards:icon": "asset:cards-statistics.statsIcon.js",
```

marta- reviewed (commented):
Usability issue: once a value for the optional `splitVar` is specified, it can no longer be removed when editing the stat definition. The list should contain an option "None" with empty `value`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -656,6 +656,9 @@ const questionnaireStyle = theme => ({
       flexBasis: "100%",
       paddingLeft: theme.spacing(4)
     },
+    filterPlaceholder: {
+        opacity: 0.3
+    },
```
Why this change?

marta- replied:
This question still stands.

marta- commented on modules/statistics/src/main/frontend/src/Statistics/Statistic.jsx:
```diff
@@ -0,0 +1,79 @@
+    rechartsData.push({"x": key, [definition["y-label"]]: expandData(value)});
+  }
+
+  let ChartType = definition["type"] == "bar" ? BarChart : LineChart;
+  let bar = definition["type"] == "bar" ? <Bar dataKey={definition["y-label"]} fill="#8884d8" /> : <Line dataKey={definition["y-label"]} type="monotone" stroke="#8884d8" />
```
A component manager would be appropriate here.

marta- replied:
Out of scope for this PR, will be addressed separately as https://phenotips.atlassian.net/browse/CARDS-1397 .

marta- commented on modules/statistics/src/main/frontend/webpack.config.js:
```diff
@@ -0,0 +1,37 @@
+module.exports = {
+  mode: 'development',
+  entry: {
+    [module_name + 'subjectsIcon']: '@material-ui/icons/AssignmentInd.js',
+    [module_name + 'subjectTypeIcon']: '@material-ui/icons/Category.js',
```
```suggestion
    [module_name + 'statsIcon']: '@material-ui/icons/BarChart.js',
```

marta- reviewed (commented):
Wrong icon asset names, and wrong icons used. The changes proposed below have not been tested.

OrdiNeu committed:
LFS-500: As a user, I can access statistics on a dashboard

Pass on the bar chart, remove an unnecessary JSON stringify/parse call

OrdiNeu committed:
LFS-500: As a user, I can access statistics on a dashboard

Fix a race condition in the statistics query servlet

marta- commented:
Some ideas for prettier charts:
* If there's no data to generate the chart with, omit the chart completely (wrapper `<Grid item>` included) instead of showing an empty chart
* Improve layout: I think the stat title would work better as a Title to the wrapper card, than under the chart as it is now.
* Clarify labels. Right now we're displaying "Patient" where we should display "Number of patients" or at least "Patients" if we want to keep it short.
* Improve colors. The purple which is default for recharts just doesn't fit in. I would make the first default color our theme's primary color (#003366), the next our theme's secondary color (#f94900) and after that we can probably use the recharts colors.
* Improve spacing: right now there's overlap between the `<Grid container>` and the search bar at the top of the page.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix support for split variables, add y-axis

marta- commented on modules/statistics/src/main/frontend/src/Statistics/Statistic.jsx:
```diff
@@ -23,38 +23,53 @@ import {
+  let chartColours = DEFAULT_PALETTE.slice();
+  let numKeys = Object.keys(allFields).length;
+  if (numKeys > DEFAULT_PALETTE.length) {
+    chartColours = chartColours.concat(palette("rainbow", numKeys-DEFAULT_PALETTE.length).map((col) => "#" + col));
```
The first color that came from `palette` doesn't look different enough from `theme.palette.secondary`:

![image](https://user-images.githubusercontent.com/651980/131882904-3f7b282b-382c-47c4-9e26-566151bd9473.png)

EDIT: Let's get rid of `theme.palette.secondary`, and use `theme.palette.primary` followed by the rainbow colors.

OrdiNeu commented:
New testing .csv that contains Gender for split field testing: 
[statisticsTest.csv](https://github.com/data-team-uhn/cards/files/7100647/statisticsTest.csv)

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix regression that preventing non-split variables from working

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Add a title to each card

marta- commented on modules/statistics/src/main/frontend/src/Statistics/Statistic.jsx:
```diff
@@ -0,0 +1,102 @@
+          <Legend />
+          {Object.keys(allFields).map((field, idx) =>
+            isBar ?
+              <Bar dataKey={field} fill={chartColours[idx]} key={idx}/>
+            :
+              <Line dataKey={field} type="monotone" stroke={chartColours[idx]} key={idx} />
```
JS console warning:

> Warning: Encountered two children with the same key, `[object Object]`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version

marta- commented on modules/statistics/src/main/frontend/src/Statistics/UserStatistics.jsx:
```diff
@@ -0,0 +1,121 @@
+    <React.Fragment>
+      <Grid container spacing={3}>
+        {currentStatistic && currentStatistic.map((stat) => {
+          return <Statistic definition={stat} key={stat} />
```
JS console warning:

> Warning: Encountered two children with the same key, `[object Object]`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

marta- commented on modules/statistics/src/main/frontend/src/Statistics/UserStatistics.jsx:
```diff
@@ -0,0 +1,121 @@
+  return (
+    <React.Fragment>
+      <Grid container spacing={3}>
+        {currentStatistic && currentStatistic.map((stat) => {
+          return <Statistic definition={stat} key={stat} />
+        })}
```
JS console warning:

> Warning: Encountered two children with the same key, `[object Object]`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

```suggestion
        {currentStatistic && currentStatistic.map((stat, i) => {
          return <Statistic definition={stat} key={i} />
        })}
```

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix console warning regarding dyplucate `key`s

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard

Update icon for stats dashboard and stats admin

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard
* More generic message when there are no stats to display
* Betterplacement of the message
* Added a top margin to the stats container as a quick fix for overlap with the global search bar

marta- commented:
Outstanding issues:

**Stats admin - usability issue**:
- Once a value for the optional splitVar is specified, it can no longer be removed when editing the stat definition. The list should contain an option "None" with empty value

**Stats user dashboard - readability issues**:
- The right part of the chars is often truncated because of the hard coded width. Display each chart inside a [ResponsiveContainer](https://recharts.org/en-US/api/ResponsiveContainer) instead.
- When a split var is used, display the labels, not the internally used values for each group. E.g. for `splitVar=Genter` - display Male, Female, Other instead of M, F, O. Additionally, these labels should be displayed in the legent in the order appear in the form.
- The `order` field specified in each stat widget does not seem to be taken into account when the stats are displayed on the user dashboard.

**Code style**: 
- A component manager would be appropriate for deciding what kind of chart to display. This can be done in the next iteration or when we start adding more chart types.

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard
* Improved the card header of each stats card: smaller title font, added an avatar dependent on chart type
* Added a message when no data is available to generate a statisticm, to replace the empty chart
* Re-positioned the legend
* Quick fix to make the y axis label clearer (used plural)

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Add the labels to the split answers

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Remove unused code

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Code cleanup

OrdiNeu committed:
LFS-500: As a user, I can access statistics on a dashboard

Allow the user to remove options in a reference input
Make the user stats dashboard obey order

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Move everything into a ResponsiveContainer

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix an issue with the split variable not counting when the question is located on a
form of a different subject type than the x variable.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix aggregation of split counts when split and x var are in different
forms

OrdiNeu commented:
New test data:
[statisticsTest.csv](https://github.com/data-team-uhn/cards/files/7137731/statisticsTest.csv)
Upload with `curl -u admin:admin "localhost:8080/Forms/" -X POST -F ":data=@statisticsTest.csv" -F":questionnaire=/Questionnaires/Patient information" -F ":subjectType=/SubjectTypes/Patient"`

[statisticsTestTumor.csv](https://github.com/data-team-uhn/cards/files/7137733/statisticsTestTumor.csv)
Upload with `curl -u admin:admin "localhost:8080/Forms/" -X POST -F ":data=@statisticsTestTumor.csv" -F":questionnaire=/Questionnaires/Tumors" -F ":subjectType=/SubjectTypes/Patient" -F ":subjectType=/SubjectTypes/Patient/Tumor"`

Both should be uploaded. The age of onset graphs should now look like this:
![image](https://user-images.githubusercontent.com/4656440/132713803-c4b37cca-2283-407e-928b-cec50fbf8fc7.png)

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix the order not being passed

OrdiNeu committed:
CARDS-1261: Administration' s Statistics: The Y and Split var and should
be depended on the selection made in X var field

Fix the bugs introduced by the merge

sashaandjic reviewed (approved):
**EDITED : found problems**
Previous comment was "Looks good."

sashaandjic commented:
Why `Y var` isn't shown in Edit statistics?
![image](https://user-images.githubusercontent.com/23243244/132773932-dde90ee0-a87e-4641-980e-59784d77e587.png)
![image](https://user-images.githubusercontent.com/23243244/132774010-00715fec-fe6d-446e-ae4c-13a4cd99a4f4.png)

![image](https://user-images.githubusercontent.com/23243244/132774051-5cfd4655-67be-4e9a-b2ce-cee09c0342c5.png)

sashaandjic commented:
For the `Tumor Location Split by Sex` the `Y var` in the live table is `Patient`, but it is blank in the Edit Statistics dialog. However when I click on the arrow it only gives `Tumor` for the `Y var` and once selected it cannot be removed. 
![image](https://user-images.githubusercontent.com/23243244/132775007-52b15c71-dc39-4822-8de2-7ee81edc50d2.png)
Modifying `Y var` in the `Age of Onset Split By Sex` breaks Statistics with `SyntaxError: Unexpected end of JSON input`.

sashaandjic commented:
The numbers in these 2 graphs should add up.
![image](https://user-images.githubusercontent.com/23243244/132798505-2476af0f-6376-47bc-8737-2b35f4fe3003.png)

OrdiNeu committed:
CARDS-1261: Administration' s Statistics: The Y and Split var and should
be depended on the selection made in X var field

Allow parents of the requiredSubjectType to be selected.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix the behaviour of split variables not being counted when they
aggregate to the same patient as a tumour

sashaandjic commented:
Undefined seems to be not accounted for. In the graph below, the patient with no gender has 3 tumors ( age 1, 3 and 8)
![image](https://user-images.githubusercontent.com/23243244/132931882-71b2bb5d-2796-469d-b7b0-bda50c726cac.png)
![image](https://user-images.githubusercontent.com/23243244/132931907-596e9e2b-134e-4e9b-a3fc-a8a54a92d7e0.png)

Another thing that I noticed earlier. The problem does no appear in the dev branch. Steps to reproduce:
- create a new patient
- create a tumor form for the patient (who does not have any tumors)
- this happens
![image](https://user-images.githubusercontent.com/23243244/132931950-829a2b8d-5022-4067-b865-a5ec779663ff.png)

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix an issue with undefined split variables not being recorded properly

marta- commented on modules/statistics/src/main/frontend/src/Statistics/Statistic.jsx:
```diff
@@ -109,11 +109,12 @@ function Statistic(props) {
-            {Object.keys(allFields).map((field, idx) =>
-              isBar ?
+            {Object.keys(allFields).sort((option1, option2) => (option1.defaultOrder - option2.defaultOrder)).map((field, idx) => {
+              console.log(field);
```
Remove leftover debug code

marta- commented:
> * create a new patient
> * create a tumor form for the patient (who does not have any tumors)
> * this happens (please note it is just a display thing, as I can proceed with the tumor creation)

Can be replicated in dev. Reported as CARDS-1393.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Remove debugging code

marta- commented:
> * When a split var is used, display the labels, not the internally used values for each group. E.g. for `splitVar=Genter` - display Male, Female, Other instead of M, F, O. Additionally, these labels should be displayed in the legend in the order appear in the form.

This comment doesn't appear to have been addressed, I'm still seeing the order `Female, Male` in the chart legend, while in the Patient Information form the order is `Male, Female`:

![image](https://user-images.githubusercontent.com/651980/133346728-d5661533-6ace-40c6-ba71-8a75947505b0.png)

For bar charts, we should display the legend only if there's a split variable. Right now we have "Patients" on the yAxis and "Patient" in the legend, which is both redundant and confusing.

![image](https://user-images.githubusercontent.com/651980/133346691-226215a9-445d-4111-9579-e024372991f5.png)

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Remove the legend when there is no split variable

sashaandjic commented:
Looking really good! Just a few small things:
- sometimes Male/Female legends are swapped - below I made 2 identical graphs where you can see that `Female` is red in one and green in the other
- sometimes the bar is not centered (see the last graph - should be centered on 5)
- "U" in "undefined"; Should the proper order be `Male`, `Female`, `Other`, `Unknown`, `Undefined`?

![image](https://user-images.githubusercontent.com/23243244/133549614-deed8c0e-5133-4236-80a6-b2920c0add78.png)
![image](https://user-images.githubusercontent.com/23243244/133549661-1638b4a9-2e7e-47e2-86c5-dc40521d6a13.png)

sashaandjic commented:
Can we do better formatting for the dates?
![image](https://user-images.githubusercontent.com/23243244/133550422-22423773-75cb-4c7d-876e-6f1d53cf4e3c.png)

sashaandjic commented:
Ahhh, bug...
When I select Tumor as `Y var`, `Split var` should only be applicable to tumors. This can be prevented at SAVE time. The graph below will break the Stats page with `SyntaxError: Unexpected end of JSON input`.
![image](https://user-images.githubusercontent.com/23243244/133551036-9b369ab7-43f4-4c86-b18a-9cdf52c650cf.png)
![image](https://user-images.githubusercontent.com/23243244/133551154-a3310a69-f6d8-4262-9544-c223f4c35810.png)

marta- commented:
> * sometimes the bar is not centered (see the last graph - should be centered on 5)
 
 There are 5 bars in each group. The group is always centered, with the green bar in the middle. But if some bars have height `0` as is the case in your screenshot it appears off. No changes needed.
 
 > * "U" in "undefined"
 
 +1 to capitalize Undefined
 
 > Should the proper order be `Male`, `Female`, `Other`, `Unknown`, `Undefined`?
 
 Yes, known issue, afaik in progress.
 
 > Should the name of the graph be `"split by sex"` or `"split by gender"`?
 
 +1 to change the labels of the statistics. Also I would get rid of the title case (use "Age of onset" instead of "Age of Onset").
 
 > Can we do better formatting for the dates?
 
 The date format from the question definition should be used.

marta- commented:
> Ahhh, bug...
> When I select Tumor as `Y var`, `Split var` should only be applicable to tumors. This can be prevented at SAVE time. The graph below will break the Stats page with `SyntaxError: Unexpected end of JSON input`.

Because we have control over stats definition for now, we can merge with this known issue and fix it separately after.

Only split var label ordering and maybe date display need to be fixed before merge. Date display can be as simple as trimming off everything after the day, for now.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Code refactor

marta- commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/Statistics/AgeOfOnset.xml:
```diff
@@ -0,0 +1,48 @@
+	<primaryNodeType>cards:Statistic</primaryNodeType>
+	<property>
+		<name>name</name>
+		<value>Age of Onset</value>
```
```suggestion
		<value>Age of onset</value>
```

marta- commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/Statistics/AgeOfOnsetSplit.xml:
```diff
@@ -0,0 +1,53 @@
+	<primaryNodeType>cards:Statistic</primaryNodeType>
+	<property>
+		<name>name</name>
+		<value>Age of Onset Split By Sex</value>
```
```suggestion
		<value>Age of onset by gender</value>
```

marta- commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/Statistics/TumorLocation.xml:
```diff
@@ -0,0 +1,48 @@
+	<primaryNodeType>cards:Statistic</primaryNodeType>
+	<property>
+		<name>name</name>
+		<value>Tumor Location</value>
```
```suggestion
		<value>Tumor location</value>
```

marta- commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/Statistics/TumorLocationSplit.xml:
```diff
@@ -0,0 +1,53 @@
+	<primaryNodeType>cards:Statistic</primaryNodeType>
+	<property>
+		<name>name</name>
+		<value>Tumor Location Split By Sex</value>
```
```suggestion
		<value>Tumor location by gender</value>
```

marta- commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,505 @@
+            Node xAnswer = xVar.adaptTo(Node.class);
+            String xValue = xAnswer.getProperty(VALUE_PROP).getString();
+
+            String splitLabel = "undefined";
```
```suggestion
            String splitLabel = "Undefined";
```

marta- committed:
CARDS-500 / CARDS-509: Sample stats for LFS

Adjusted the stats labels, capitalized "Undefined" label corresponding to data points without a value for the split variable.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Reorder the options according to their order in the questionnaire.

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Remove unused style

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Add proper date formatting
Remove unused styling

sashaandjic commented:
Date of birth seems to be shifted by 1. All of the patients have DOB from Sep1 to Sep4. Also, the dates on the X axis are not sorted.
Please take a look at the bottom graphs.
![image](https://user-images.githubusercontent.com/23243244/133967848-52d88e7b-d595-49e0-aa0d-59438749267d.png)

The other thing that I noticed is this:
![image](https://user-images.githubusercontent.com/23243244/133967916-cf4166ca-f0c9-4e47-ab28-712cbcdf21c5.png)

marta- commented:
> The other thing that I noticed is this:
> ![image](https://user-images.githubusercontent.com/23243244/133967916-cf4166ca-f0c9-4e47-ab28-712cbcdf21c5.png)

This was introduced  by PR #775 and is fixed in PR #778.

marta- commented:
> Date of birth seems to be shifted by 1. All of the patients have DOB from Sep1 to Sep4.

@sashaandjic I remember you having date shifting issues because of your vm settings. If nobody else can replicate this, I would call it a non-issue for now.

> Also, the dates on the X axis are not sorted.

@OrdiNeu This needs to be fixed.

sashaandjic commented:
> > Date of birth seems to be shifted by 1. All of the patients have DOB from Sep1 to Sep4.
> 
> @sashaandjic I remember you having date shifting issues because of WSL. If nobody else can replicate this, I would call it a non-issue for now.

Ah, yes. WSL has problem with dates.

acrowthe commented:
Reproduced DOB issues. All DOBs were entered as September 1st and September 2nd:
![image](https://user-images.githubusercontent.com/20056751/134048514-074651e5-e05f-4819-82cb-f5300ff4ce5a.png)

Steps followed:
1. Have time zone set to UTC-4 (Eastern daylight time during the summer).
2. Set the DOB to September 1st for 4 subject and September 2nd for 4 subjects (time of day when the data is entered is irrelevant, tested 00:30, 01:30, 10:30 and 23:30
3. Set the time zone to UTC-5 (Eastern Daylight time during the winter, or CDT during the summer)
4. Restart sling as changing the time backwards crashed sling
4. Observe that all previously entered DOBs are showing up as August 31 and September 1st, rather than September 1st and 2nd
5. Create a new DOB on September 1st and September 2nd, observing that they are displayed on the correct date

acrowthe commented:
Looks like the issue is only reproducible when manually changing timezones from Eastern to Central time zones, not when changing the date to trigger daylight savings time.
- Changing date from September to December is fine, despite automatic time zone change from EDT to EST
- Changing time zone from EST to CST manually results in all statistic DOBs being 1 day early, even if entered after time zone change

Simple date issue reproduction steps:
1. Using the admin dashboard, create a statistic as seen here:
![image](https://user-images.githubusercontent.com/20056751/134050283-965456bd-9e20-428d-8a0e-303c3a7a45dc.png)
2. Have timezone set to EDT with the date in September, so timezone is EDT or UTC-4.
3. Enter a DOB of September 1st on a patient information form. Observe that the statistic also has that date as September 1st.
4. Change the date to December, automatically changing timezones to EST or UTC-5. Observe that statistics still show September 1st.
5. Restart Sling. Observe statistic is still September 1st.
6. Manually change the timezone from EST (UTC-5) to CST (UTC-6) and date to the next day to avoid sling crashing. Observe statistic now shows August 31st instead of September 1st.
7. Create a new date of birth on September 3rd. Observe that statistic shows that the date of birth is September 2nd, not September 3rd as entered

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Sort the x axis

sashaandjic reviewed (approved):
tested

marta- commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,505 @@
+
+    private static final String VALUE_PROP = "value";
+
+    @SuppressWarnings({"checkstyle:ExecutableStatementCount", "checkstyle:CyclomaticComplexity", "checkstyle:JavaNCSS"})
```
Would this `@SuppressWarnings` still be necessary if we moved the parsing logic (lines 76-99) to a separate function?

marta- commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,505 @@
+     * @param counts Map of {SubjectID, {Split variable label, count}}
+     * @return map of {x var, {split var, count}}
+     */
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity"})
```
Would this `@SuppressWarnings` still be necessary if we moved the logic for obtaining `splitLabel` (lines 263-272) to a separate function?

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Code refactor to remove code smell

sdumitriu commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -284,29 +272,37 @@ protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse
         } catch (PathNotFoundException e) {
             LOGGER.error("Value does not exist for question: {}", e.getMessage(), e);
+        } finally {
+            this.splitLabels.remove();
```
Since there is no loop in this method, only one call to `getSplitLabels` will be made during the execution of this method. Thus, I don't think there's a need to even store `splitLabels` in a `ThreadLocal` as it is. Either move the `remove` call into `addDataSplit`, right after the `for`, or don't bother using a `ThreadLocal` at all.

OrdiNeu replied:
Oops, fixed in `d82847653` -- it was meant to be used outside of the loops in `addDataSplit`

OrdiNeu committed:
CARDS-500: As a user, I can access statistics on a dashboard

Fix the misplaced .remove() call

sashaandjic reviewed (approved):
re-tested

marta- commented:
> Question: Can we make more room between X axis and its description?

@OrdiNeu According to r[echarts documentation & examples](https://recharts.org/en-US/api/Label), it would appear that an `offset` of about -10 o the X axis label would slide the label down enough to make everything more readable, can you give it a try?

marta- commented on modules/statistics/src/main/frontend/src/Statistics/Statistic.jsx:
```diff
@@ -0,0 +1,171 @@
+          <ChartType data={rechartsData}>
+            <CartesianGrid strokeDasharray="3 3" />
+            <XAxis dataKey="x">
+              <Label value={definition["x-label"]} offset={0} position="insideBottom" />
```
```suggestion
              <Label value={definition["x-label"]} offset={-10} position="insideBottom" />
```

marta- commented on modules/statistics/src/main/frontend/src/Statistics/Statistic.jsx:
```diff
@@ -0,0 +1,171 @@
+            <XAxis dataKey="x">
+              <Label value={definition["x-label"]} offset={0} position="insideBottom" />
+            </XAxis>
+            <YAxis allowDecimals={false} label={{ value: yAxisLabel, angle: -90, position: 'insideLeft' }} />
```
We could probably also bring the Y label closer to the Y axis to make everything more balanced:
```suggestion
            <YAxis allowDecimals={false} label={{ value: yAxisLabel, angle: -90, position: 'insideLeft', offset: 10 }} />
```

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard

Addressing comments: better positioning of axis labels

marta- commented:
The last 2 commits improve general spacing and alignment in the stats box (labels, legend, distance between box title and chart, consistent heights of charts with and without legend).

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard
Better positioning of the legend, improved spacing and alignment within each Stats box

sdumitriu commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/Statistics/AgeOfOnsetSplit.xml:
```diff
@@ -0,0 +1,53 @@
+<!--
```
I would use a more explicit name like `AgeOfOnsetByGender` (or `AgeOfOnsetSplitByGender`).

sdumitriu commented on lfs-resources/clinical-data/src/main/resources/SLING-INF/content/Statistics/TumorLocationSplit.xml:
```diff
@@ -0,0 +1,53 @@
+<!--
```
Same here, `TumorLocationByGender`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -41,13 +41,13 @@ const ALL_QUESTIONNAIRES_URL = "/Questionnaires.deep.json";
+  const { classes, disabled, onChangeFilters, questionnaire, filtersJsonString, statisticFilters, parentHandler, statisticFiltersValue } = props;
   // Filters, as displayed in the dialog, and filters as actually saved
-  const [editingFilters, setEditingFilters] = useState([]);
+  const [editingFilters, setEditingFilters] = useState([]); // TODO: existing
```
Is the loading of filters from the URL, which we did recently implement, considered a fix for this TODO?

```suggestion
  const [editingFilters, setEditingFilters] = useState([]);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -416,6 +417,46 @@ function Filters(props) {
+    if (!filterRequestSent) {
+      grabFilters(FILTER_URL);
+    }
+
```
I think this is not needed, we always grab filters in an `useEffect` right at the top, don't we?

```suggestion
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Fields.jsx:
```diff
@@ -91,10 +94,15 @@ let Fields = (props) => {
+        edit ?
+          Object.entries(JSON).filter(([key, _]) => key != "//REQUIRED").map(([key, value]) => displayEditField(key, value))
+        :
+          Object.keys(getAllKeys(JSON)).filter(([key, _]) => key != "//REQUIRED").map(key => (data[key] ? displayStaticField(key) : ''))
```
Should this be made a bit more generic and exclude anything starting with `//`?

```suggestion
          Object.entries(JSON).filter(([key, _]) => !key.startsWith("//")).map(([key, value]) => displayEditField(key, value))
        :
          Object.keys(getAllKeys(JSON)).filter(([key, _]) => !key.startsWith("//")).map(key => (data[key] ? displayStaticField(key) : ''))
```

Also, the last line is buggy, it should probably be:

```suggestion
          Object.entries(JSON).filter(([key, _]) => !key.startsWith("//")).map(([key, value]) => displayEditField(key, value))
        :
          Object.keys(getAllKeys(JSON)).filter((key) => !key.startsWith("//")).map(key => (data[key] ? displayStaticField(key) : ''))
```

But even like this, I'm not sure it will work properly since `displayStaticField` kind of requires a value, which is no longer passed. The only consequence so far seems to be that markdown fields are not going to be rendered: in the questionnaire editor, edit any question, add a formatted description, save -> the description is not formatted in the questionnaire editor.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Fields.jsx:
```diff
@@ -32,6 +32,9 @@ import ObjectInput from "./ObjectInput";
 import MarkdownTextField from "./MarkdownTextField";
 import FormattedText from "../components/FormattedText.jsx";
+import MDEditor from '@uiw/react-md-editor';
+import ReferenceInput from "./ReferenceInput";
```
These don't seem to be used:

```suggestion
```

sdumitriu commented on modules/data-entry/src/main/provisioning/model.txt:
```diff
@@ -49,6 +49,7 @@
             create path (cards:FormsHomepage) /Forms \
             create path (cards:SubjectsHomepage) /Subjects \
             create path (cards:SubjectTypesHomepage) /SubjectTypes \
+            create path (cards:StatisticsHomepage) /Statistics \
```
These changes don't fit here, but duplicating all the runmodes in the statistics module is not that good either. It can be like this for now while we look for a better way to set up different permission schemes.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/EntityIdentifier.jsx:
```diff
@@ -35,6 +35,8 @@ export function getEntityIdentifier(row) {
       case "cards:SubjectType":
         return row.label;
+      case "cards:Statistic":
+        return row.name;
```
We have too many names for names...

sdumitriu commented on modules/statistics/src/main/frontend/package.json:
```diff
@@ -0,0 +1,56 @@
+{
+    "name": "cards-statistics",
+    "version": "0.1.0",
```
```suggestion
    "version": "0.9-SNAPSHOT",
```

But maybe we should have this as `${project.version}`...

sdumitriu committed:
[cleanup] Fixed indentation

sdumitriu committed:
[misc] Updated and cleaned up dependencies

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+import { fetchWithReLogin, GlobalLoginContext } from "../login/loginDialogue.js";
+
+/**
+ * Createa the LIveTable cell contents for a given node. This generates a link to the
```
```suggestion
 * Create the LiveTable cell contents for a given node. This generates a link to the
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+    return <></>;
+  }
+
+  if (node["jcr:primaryType"] == "lfs:Question") {
```
All the prefixes should be `cards:` instead of `lfs:`, does this actually work?

Nope, it doesn't.

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Fixed deprecated lfs: namespace

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+    {
+      "key": "name",
+      "label": "Name",
+      "format": "string",
```
In the future there should be a link here. Right now the only place where it could lead is to the edit mode, but maybe there should be a view mode for a statistic?

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+                              entryPath={row["@path"]}
+                              entryName={row.label}
+                              onComplete={dialogSuccess}
+                              entryType={"Subject Type"}
```
Maybe this?

```suggestion
                              entryType={"Statistic"}
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+                              entryName={row.label}
+                              onComplete={dialogSuccess}
+                              entryType={"Subject Type"}
+                              admin={true}
```
This is no longer used, is it?

```suggestion
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+ *
+ * @param {Object} node The lfs:SubjectType or lfs:Question node to generate a link for
+ */
+function CreateTableCell(node) {
```
This is not used as a React component, so I think it should not start with a capital letter:

```suggestion
function createTableCell(node) {
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+    {
+      "key": "xVar",
+      "label": "X-axis",
+      "format": (stat) => CreateTableCell(stat?.xVar),
```
```suggestion
      "format": (stat) => createTableCell(stat?.xVar),
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+    {
+      "key": "yVar",
+      "label": "Y-axis",
+      "format": (stat) => CreateTableCell(stat?.yVar),
```
```suggestion
      "format": (stat) => createTableCell(stat?.yVar),
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+    {
+      "key": "splitVar",
+      "label": "Split",
+      "format": (stat) => CreateTableCell(stat?.splitVar),
```
```suggestion
      "format": (stat) => createTableCell(stat?.splitVar),
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+      "format": (row) => (<>
+                            <DeleteButton
+                              entryPath={row["@path"]}
+                              entryName={row.label}
```
```suggestion
                              entryName={row.name}
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+      <Card>
+       <CardHeader
+        title={
+          <Button className={classes.cardHeaderButton}>
```
Can someone remind me why are the headers buttons?

marta- replied:
They were buttons in the MaterialDashboardReact templates that we used initially and had some cool styling applied to then. They don't need to be buttons anymore since we're no longer using that part of the MDR style.

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+          <NewItemComponent
+            title="Create new statistic"
+            onClick={() => {setDialogOpen(true); setNewStat(true); setCurrentId();}}
+            inProgress={false}
```
```suggestion
            inProgress={dialogOpen}
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+  withStyles,
+} from "@material-ui/core";
+import statisticsStyle from "./statisticsStyle.jsx";
+import NewItemComponent from "../components/NewItemButton.jsx";
```
Why is this renamed?

```suggestion
import NewItemButton from "../components/NewItemButton.jsx";
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+          </Button>
+        }
+        action={
+          <NewItemComponent
```
```suggestion
          <NewItemButton
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -0,0 +1,311 @@
+        }
+        classes={{
+          action: classes.newFormButtonHeader
+        }}
```
Is this needed, or just deprecated code?

```suggestion
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/Filters.jsx:
```diff
@@ -59,6 +59,7 @@ function Filters(props) {
   const [filterRequestSent, setFilterRequestSent] = useState(false);
   const [toFocus, setFocusRow] = useState(null);
   const notesComparator = "notes contain";
+  const [statisticValue, setStatisticValue] = useState();
```
I don't like this very explicit cross-dependency between filters and statistics.

sdumitriu replied:
Huh, actually I think NONE of the changes in this file are needed, the functionality is provided by `ReferenceInput`, which doesn't depend on `Filters`.

sdumitriu commented on modules/statistics/src/main/resources/SLING-INF/nodetypes/statistics.cnd:
```diff
@@ -0,0 +1,81 @@
+
+//-----------------------------------------------------------------------------
+// Statistic: a group of data to be visually displayed
+[cards:Statistic] > sling:Folder, mix:referenceable, mix:versionable
```
Do they need to be versionable? With this set, there are a few issues:
- predefined statistics (loaded from source XMLs) are always checked out, so they only have one version
- new statistics (defined in the administration) are always checked in, so they get a new version each time we save them
- there's no `mix:lastModified`, so we don't keep track of last modified date

I think for the moment we don't need to worry about versioning these, we'll do a full implementation later if we decide we need to.

```suggestion
[cards:Statistic] > sling:Folder, mix:referenceable
```

sdumitriu commented on modules/statistics/src/main/resources/SLING-INF/nodetypes/statistics.cnd:
```diff
@@ -0,0 +1,81 @@
+
+  // A reference to the x-axis variable (question or section)
+  // No full text search, since it's just a non-textual reference.
+  - xVar (reference) mandatory multiple
```
Why `multiple`? Do we support multiple variables? I think this is better:

```suggestion
  - xVar (reference) mandatory nofulltext
```

sdumitriu commented on modules/statistics/src/main/resources/SLING-INF/nodetypes/statistics.cnd:
```diff
@@ -0,0 +1,81 @@
+
+  // A reference to the y-axis variable (subjectType)
+  // No full text search, since it's just a non-textual reference.
+  - yVar (reference) mandatory multiple
```
```suggestion
  - yVar (reference) mandatory nofulltext
```

sdumitriu commented on modules/statistics/src/main/resources/SLING-INF/nodetypes/statistics.cnd:
```diff
@@ -0,0 +1,81 @@
+  - sling:resourceSuperType (STRING) = "cards/Resource" mandatory autocreated protected
+
+  // Set a default title.
+  - title (String) = "Statistics" mandatory autocreated
```
```suggestion
  - title (String) = "Statistics" mandatory autocreated
```

sdumitriu commented on modules/statistics/src/main/resources/SLING-INF/content/Extensions/Views/UserStatistics.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "jcr:primaryType": "cards:Extension",
+    "cards:extensionPointId": "cards/coreUI/view",
+    "cards:extensionName": "User Statistics",
```
I think it's better like this:

```suggestion
    "cards:extensionName": "Statistics",
```

(this is visible in the browser window/tab title)

sdumitriu commented on modules/statistics/src/main/provisioning/model.txt:
```diff
@@ -0,0 +1,37 @@
+    create path (cards:StatisticsHomepage) /Statistics
+
+    set ACL for everyone
+        allow   jcr:read    on /content
```
```suggestion
        allow   jcr:read    on /Statistics
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/statisticsStyle.jsx:
```diff
@@ -0,0 +1,44 @@
+  under the License.
+*/
+
+// Taken from https://www.creative-tim.com/product/material-dashboard-react
```
Really?

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/statisticsStyle.jsx:
```diff
@@ -0,0 +1,44 @@
+  KIND, either express or implied.  See the License for the
+  specific language governing permissions and limitations
+  under the License.
+*/
```
This looks like an APL header, just wrongly formatted. This will fix the RAT exception:

```suggestion
//
//  Licensed to the Apache Software Foundation (ASF) under one
//  or more contributor license agreements.  See the NOTICE file
//  distributed with this work for additional information
//  regarding copyright ownership.  The ASF licenses this file
//  to you under the Apache License, Version 2.0 (the
//  "License"); you may not use this file except in compliance
//  with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the License is distributed on an
//  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
//  KIND, either express or implied.  See the License for the
//  specific language governing permissions and limitations
//  under the License.
//
```

However, if this is indeed copied from MDR, it should have a different license.

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/UserStatistics.jsx:
```diff
@@ -0,0 +1,129 @@
+  let fetchStat = (stat) => {
+    // for each existing statistic, get full.json
+    // pathnames will be sent in request body, that's why deep.json is used
+    fetchWithReLogin(globalContext, `/Statistics/${stat['@name']}.deep.json`)
```
```suggestion
    fetchWithReLogin(globalContext, `${stat['@path']}.deep.json`)
```

sdumitriu commented on modules/statistics/src/main/frontend/src/Statistics/UserStatistics.jsx:
```diff
@@ -0,0 +1,129 @@
+//
```
With a lot of data and a slow connection, this will not behave very well. Current behavior:

- user opens the stats page
- nothing is displayed
- slowly, stats start showing up one by one, jumping around as the data comes in
- no indication if another item will show up or not

Better behavior (to be improved in the future)

- user opens the stats page
- a loading animation shows up
- when the statistics definitions are loaded, we display a grid with placeholders
- each placeholder should have a title and a loading animation in it
- as data comes in, grid items are replaced with the proper chart

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+import { Input, MenuItem, Select, Typography, withStyles } from "@material-ui/core";
+
+import EditorInput from "./EditorInput";
+import LiveTableStyle from "../dataHomepage/tableStyle.jsx";
```
Weird, why does this need **livetable** styles?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+  const [ options, setOptions ] = useState([]);
+  const [ restrictions, setRestrictions ] = useState([]);
+
+  const isNumeric = value.filter == "numeric";
```
Is this ever in use?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+  }
+
+  // From an array of fields, turn it into a react component
+  let GetReactComponentFromFields = (fields, nested=false) => {
```
This is also a plain function, not a react component, so it should start with a lowercase letter.

```suggestion
  let getReactComponentFromFields = (fields, nested=false) => {
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+        // Arrays represent Questionnaires of Sections
+        // which we'll need to turn into opt groups
+        return [<MenuItem className={classes.categoryHeader} disabled>{path[0]}</MenuItem>,
+          GetReactComponentFromFields(path.slice(1), true)];
```
```suggestion
          getReactComponentFromFields(path.slice(1), true)];
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+        input={<Input id={objectKey} />}
+        renderValue={(value) => titleMap[value]}
+      >
+        {GetReactComponentFromFields(options)}
```
```suggestion
        {getReactComponentFromFields(options)}
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+        // We only care about children that are cards:Questions or cards:Sections
+        if (object["jcr:primaryType"] == "cards:Question") {
+          // If we are only interested in numeric values, ignore this one unless it is numeric
+          if (isNumeric && !(object["dataType"] == 'decimal' || object["dataType"] == 'long')) {
```
What about `double`?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+      fetchRequest = fetchWithReLogin(globalLoginDisplay, url)
+        .then((response) => response.ok ? response.json() : Promise.reject(response))
+        .then((json) => {
+          let nodePath = json["rows"][0]["@path"];
```
For safety, let's do this instead:

```suggestion
          let nodePath = json["rows"]?.[0]?.["@path"];
          nodePath || Promise.reject("Invalid reference: " + field);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx:
```diff
@@ -0,0 +1,292 @@
+let SUBJECT_TYPE_URL = "/SubjectTypes.paginate?offset=0&limit=100&req=0";
+
+// Reference Input field used by Edit dialog component
+let ReferenceInput = (props) => {
```
This is trying to be both a generic "Reference" input, and very specific to questions and subject types relevant to the selected question, failing to do both. I don't think it can be easily turned into a proper generic input, so I would separate the functionalities into `QuestionReferenceInput` and `SubjectTypeReferenceInput`.

sdumitriu commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,503 @@
+
+            boolean foundAnswer = false;
+            while (formSubject.getDepth() > 0) {
+                if (formSubject.getProperty("type").getNode().getIdentifier().equals(correctType)) {
```
This fails on `formSubject.getProperty("type")` if we don't find the correct type and go too far up the tree, since the subjects homepage does not have that property.

```suggestion
                if (formSubject.hasProperty("type")
                    && formSubject.getProperty("type").getNode().getIdentifier().equals(correctType)) {
```

sdumitriu commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,503 @@
+     * @param subjectType Subject type of the subject for the answer's form
+     * @return The filtered iterator
+     */
+    private Map<String, Map<Resource, String>> filterAnswersWithType(Map<Resource, String> data, Node subjectType)
```
I'm not sure this implementation is correct, I think the aggregation|filtering|splitting should happen in a different order, since this doesn't support splitting by a variable that applies to a parent of the targeted subject type, e.g. instead #Patients by age of onset split by gender, have #Tumors by age of onset split by gender.

sdumitriu commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,503 @@
+     * @param node A node corresponding to an cards:Answer
+     * @return A node corresponding to the parent subject type
+     */
+    public Node getParentNode(Node node) throws RepositoryException
```
A parent node is the direct parent of a node, regardless of type. I would name this `getFormNode(Node answer)`.

sdumitriu commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,503 @@
+        // If we somehow reach the top level, return an error
+        Node answerParent = node.getParent();
+        while (!"cards:Form".equals(answerParent.getPrimaryNodeType().getName()) && answerParent.getDepth() != 0) {
+            answerParent = answerParent.getParent();
```
I think this may fail for a non-admin user since `getParent()` may return `null` if the parent (root) is inaccessible. Right now we have a workaround another bug which does allow access to `/` for all users, but with the sling upgrade that workaround is no longer needed. I think we also need to check if `answerParent != null`.

sdumitriu commented on modules/statistics/src/main/java/io/uhndata/cards/statistics/StatisticQueryServlet.java:
```diff
@@ -0,0 +1,503 @@
+        if (answerParent.getDepth() == 0) {
+            String error = String.format("Tried to obtain the parent cards:Form for node {} but failed to find one",
+                node.getPath());
+            throw new RepositoryException(error);
```
I think throwing an exception here is too drastic, I think it will stop all processing, while the expected action is to just ignore this answer.

sdumitriu reviewed (commented):
Oh, another thing here: I thought we decided we'll run the queries with increased privileges, so we'll need to add a provisioning configuration to allow administrative access for this module, and change either the session, or the resource resolver, to use the service login.

sdumitriu committed:
[cleanup] Applied codestyle

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Various improvements to the statistics servlet

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Undo changes to Filters.jsx, since the functionality now resides fully in ReferenceInput

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
- Rename split stats to mention what they are split by
- Use patient's "age at first malignant Dx" instead of tumor's "age at diagnosis"

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Bugfix: markdown fields are displayed unformatted in the questionnaire editor

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Non-component functions should not start with an upper letter

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Statistics don't need to be versionable yet
xVar and yVar should not be multivalued

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
The page title should not be "User Statistics", just "Statistics"

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Fixed leftover copy/paste error in granting rights

sdumitriu committed:
[misc] Fixed license header

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Simpler way of computing the path to a questionnaire

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Stats admin fixes:
- fixed links to questions
- proper marking of the New button as in progress
- use the corrent item type in the delete dialog
- cleanup comments, imports and unnecessary CSS

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Better error handling when failing to access the referenced node

sdumitriu committed:
CARDS-500: As a user, I can access statistics on a dashboard
Bugfix: statistics table question link lead to wrong place

sashaandjic reviewed (approved):
tested - all good!

sashaandjic commented:
after introducing "None" option for gender, the Undefined counts "None" and blanks. I have 1 patient with no gender specified and one with "None". Now the question is: Should we count "None" as Undefined or should we keep them separate? 
![image](https://user-images.githubusercontent.com/23243244/136257797-6d3091e8-dcfc-46d4-bbbc-613f14ecdd85.png)

sashaandjic commented:
Not sure about this. Am I missing something?
![image](https://user-images.githubusercontent.com/23243244/136259836-945fc4a5-2749-43cb-9549-a08b4f7dec0e.png)

marta- commented:
@sashaandjic The last discovered bug was addressed in f5fd0ae .

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard
Bugfix: `ReferenceInput` no longer registered as a field displayer (bug introduced by commit a0a4e2d)

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard
Bugfix: markdown fields are displayed unformatted in the questionnaire editor
(Original fix attempt from commit a0a4e2d doesn't actually do it)

marta- committed:
[misc] Fix regression: nested fields no longer displayed in the questionnaire editor view mode
(e.g. minAnswers, maxAnswers, displayType, ...)

marta- committed:
CARDS-500 / CARDS-508 New entry in the admin screen: Statistics
Only show the circular progress around the (+) button when creating a new statistic, not when editing an existing one

marta- commented:
> Everything seems to work fine in the latest version.
> 
> One minor comment: `Undefined` seems too technical, should we use `Unspecified` or `Unset` instead?

"Unspecified" sounds good. "Not specified" sounds better.

marta- committed:
CARDS-500: As a user, I can access statistics on a dashboard
Use "Not specified" instead of "Undefined" when the value of the split variable is not set

sashaandjic reviewed (approved):
tested
